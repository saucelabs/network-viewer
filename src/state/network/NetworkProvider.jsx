import React, { useEffect, useReducer, useMemo } from 'react';
import PropTypes from 'prop-types';

import { reducer, initialState as defaultState } from './reducer';
import { updateData, fetchFile, updateScrollToIndex } from './actions';
import { NetworkContext } from './Context';
import { findRequestIndex } from '../../utils';
import { ROW_ID_PREFIX } from '../../constants';

const NetworkProvider = (props) => {
  const {
    data,
    file,
    fetchOptions,
    initialState,
    scrollTimeStamp,
    scrollRequestPosition,
    autoHighlightChange,
    onDataLoaded,
    onDataError,
    onPause,
    onResume,
    onReset,
  } = props;

  const [state, dispatch] = useReducer(reducer, initialState);
  const callbacks = {
    onPause,
    onResume,
    onReset,
  };
  const value = useMemo(() => [state, dispatch, callbacks], [state]);
  const selectedReqIndex = value[0].get('selectedReqIndex');
  const requestData = value[0].get('data');
  const showReqDetail = value[0].get('showReqDetail');
  const actualData = value[0].get('actualData');
  const error = value[0].get('error');

  // Update data onChange of network data
  useEffect(() => {
    if (data && data.log && data.log.entries) {
      updateData(dispatch)(data);
    }
  }, [data]);

  // Fetch HAR file onChange of file prop
  useEffect(() => {
    if (file) {
      fetchFile(dispatch)(file, fetchOptions);
    }
  }, [file]);

  useEffect(() => {
    if (actualData.size && onDataLoaded) {
      onDataLoaded(actualData);
    }
  }, [actualData]);

  useEffect(() => {
    if (error && onDataError) {
      onDataError(error);
    }
  }, [error]);

  // Find nearby request-rowId and update scrollIndex on scrollTimeStamp receive
  useEffect(() => {
    const shouldChangeHighlight = showReqDetail ? autoHighlightChange : true;
    if (scrollTimeStamp && shouldChangeHighlight) {
      const reqIndex = findRequestIndex({
        data: requestData,
        timestamp: scrollTimeStamp,
        position: scrollRequestPosition,
      });
      if (reqIndex || reqIndex === 0) {
        updateScrollToIndex(dispatch)(requestData.get(reqIndex));
      }
    }
  }, [scrollTimeStamp, actualData]);

  // Scroll to request row onChange of scrollToIndex
  // setTimeout is required when reqDetail is visible,
  // it allows DOM to adjust before we scroll it to highlighted request
  useEffect(() => {
    if (selectedReqIndex) {
      setTimeout(() => {
        const element = document.getElementById(ROW_ID_PREFIX + selectedReqIndex);
        if (element) {
          element.scrollIntoView({
            alignToTop: true,
            behavior: 'smooth',
          });
        }
      }, 300);
    }
  }, [selectedReqIndex]);

  return (
    <NetworkContext.Provider
      value={value}
      {...props}
    />
  );
};

NetworkProvider.propTypes = {
  autoHighlightChange: PropTypes.bool,
  data: PropTypes.object,
  fetchOptions: PropTypes.object,
  file: PropTypes.string,
  initialState: PropTypes.object,
  onDataError: PropTypes.func,
  onDataLoaded: PropTypes.func,
  onPause: PropTypes.func,
  onReset: PropTypes.func,
  onResume: PropTypes.func,
  scrollRequestPosition: PropTypes.oneOf(['before', 'after', 'near']),
  scrollTimeStamp: PropTypes.number,
};

NetworkProvider.defaultProps = {
  autoHighlightChange: false,
  data: null,
  fetchOptions: { withCredentials: true },
  file: null,
  initialState: defaultState,
  onDataError: null,
  onDataLoaded: null,
  onPause: null,
  onReset: null,
  onResume: null,
  scrollRequestPosition: 'near',
  scrollTimeStamp: null,
};

export default NetworkProvider;
