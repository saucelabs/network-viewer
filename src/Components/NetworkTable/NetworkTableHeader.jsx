import React from 'react';
import classNames from 'classnames/bind';

import Styles from './NetworkTable.styles.scss';
import { useNetwork } from '../../state/network/Context';
import { useTheme } from '../../state/theme/Context';
import { getViewerFields } from '../../utils';

const context = classNames.bind(Styles);

const NetworkTableHeader = () => {
  const { state } = useNetwork();
  const showReqDetail = state.get('showReqDetail');
  const tableHeaderWidth = state.get('tableHeaderWidth');
  const { showWaterfall } = useTheme();

  const columns = getViewerFields(showReqDetail, showWaterfall);

  return (
    <div
      className={Styles['network-table-header']}
      style={{ width: tableHeaderWidth }}
    >
      {Object.entries(columns)
        .map(([datakey, {
          key,
          name,
        }]) => (
          <div
            key={key}
            className={context(
              'table-column',
              'value-cell',
              datakey,
              { 'limited-cols': showReqDetail },
              { 'show-waterfall': showWaterfall },
            )}
          >
            {name}
          </div>
        ))}
    </div>
  );
};

export default NetworkTableHeader;
