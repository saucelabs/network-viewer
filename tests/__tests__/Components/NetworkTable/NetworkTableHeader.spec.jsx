import React from 'react';
import { mount } from 'enzyme';

import NetworkTableHeader from './../../../../src/Components/NetworkTable/NetworkTableHeader';
import NetworkProvider from '../../../../src/state/network/NetworkProvider';

describe('NetworkTableHeader', () => {
  it('renders without crashing', () => {
    const element = mount(
      <NetworkProvider>
        <NetworkTableHeader />
      </NetworkProvider>,
    );
    expect(element).toMatchSnapshot();
  });
});
