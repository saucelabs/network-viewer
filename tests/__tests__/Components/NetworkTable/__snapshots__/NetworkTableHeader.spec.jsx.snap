// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkTableHeader renders without crashing 1`] = `
<NetworkProvider
  autoHighlightChange={false}
  data={null}
  fetchOptions={
    Object {
      "withCredentials": true,
    }
  }
  file={null}
  initialState={
    Immutable.Map {
      "search": Object {
        "name": "URL",
        "value": "",
      },
      "error": null,
      "statusFilter": Object {
        "name": "All",
        "value": null,
      },
      "scrollToIndex": null,
      "data": Immutable.List [],
      "tableHeaderWidth": "100%",
      "selectedReqIndex": null,
      "reqDetail": null,
      "showReqDetail": false,
      "typeFilter": Object {
        "name": null,
        "value": null,
      },
      "rawData": Object {
        "log": Object {
          "creator": Object {
            "name": "",
            "version": "",
          },
          "entries": Array [],
          "pages": Array [],
          "version": "",
        },
      },
      "totalNetworkTime": null,
      "actualData": Immutable.List [],
      "dataSummary": Immutable.Map {},
      "loading": false,
      "sort": Object {
        "isAsc": true,
        "key": "startedDateTime",
      },
    }
  }
  onDataError={null}
  onDataLoaded={null}
  onPause={null}
  onRequestSelect={null}
  onReset={null}
  onResume={null}
  scrollRequestPosition="near"
  scrollTimeStamp={null}
>
  <NetworkTableHeader>
    <div
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <div
        className="table-column value-cell file show-waterfall"
        key="filename"
      >
        Path
      </div>
      <div
        className="table-column value-cell domain show-waterfall"
        key="domain"
      >
        Domain
      </div>
      <div
        className="table-column value-cell status show-waterfall"
        key="status"
      >
        Status
      </div>
      <div
        className="table-column value-cell method show-waterfall"
        key="method"
      >
        Method
      </div>
      <div
        className="table-column value-cell type show-waterfall"
        key="type"
      >
        Type
      </div>
      <div
        className="table-column value-cell size show-waterfall"
        key="size"
      >
        Size
      </div>
      <div
        className="table-column value-cell time show-waterfall"
        key="time"
      >
        Time
      </div>
      <div
        className="table-column value-cell waterfall show-waterfall"
        key="waterfall"
      >
        Waterfall
      </div>
    </div>
  </NetworkTableHeader>
</NetworkProvider>
`;
